### PR TITLE
User controllable serial port power for TX16S

### PR DESCRIPTION
**Pull request by rotorman**
Solves #1766 by adding checkboxes into UI, for the user to explicitly enable or disable the power output on supply pins next to serial ports available on some radios (presently only TX16S has this feature).

Two example screenshots from physical TX16S mkII:

grafik

grafik

As USB VCP has no controllable power, no check box is presented for it.

The augmented serial port section in `radio.yml`, corresponding to the last screenshot:
```
serialPort: 
   AUX1:
      mode: SBUS_TRAINER
      power: 0
   AUX2:
      mode: GPS
      power: 1
   VCP:
      mode: CLI
      power: 0
```

Internally, the `g_eeGeneral.serialPort` was expanded with this PR to 8-bits per serial port, lower 4 bits are used for mode (0 = OFF, 1 = Telemetry Mirror, 2= Telemetry In, 3 = SBUS Trainer, 4 = LUA, 5 = CLI, 6 = GPS), ~~5th bit~~ 8th bit for power state (0 = OFF, 1 = ON), the bits 5 to 7 remain presently unused.
**Update:** the power state bit was in a later commit changed to the most-significant bit due to suggestion by @pfeerick

Thx to @raphaelcoeffic for numerous tips with YAML.

Needs Companion support - @elecpower yelp!

Should we limit this feature presently to TX16S target only?

### CODE REVIEW

File: radio/src/gui/colorlcd/radio_hardware.cpp [hunk 1]
```diff
@@ -235,6 +235,20 @@ class SerialConfigWindow : public FormGroup
       aux->setAvailableHandler(
           [=](int value) { return isSerialModeAvailable(port_nr, value); });
       grid.nextLine();
+
+      if (port_nr < SP_VCP)
+      {
+          new StaticText(this, grid.getLabelSlot(), "Port power:", 0, COLOR_THEME_PRIMARY1);
```
Comment (pfeerick) [inline radio/src/gui/colorlcd/radio_hardware.cpp:241]: grid.getLabelSlot(true) will indent the text ;)

**Comment by pfeerick** [comment]
Is there any reason for the power bit to not be the highest, so that it isn't in the middle of the array if more options are added later? Not that it really makes any difference either way. 

I think the question should more be "what other transmitters support soft power control of the serial power pins... if the answer is none... the answer is pretty clear ;)

File: radio/src/gui/colorlcd/radio_hardware.cpp (hunk 1 above)
Comment (rotorman) [reply radio/src/gui/colorlcd/radio_hardware.cpp]: Thanks, updated!

**Comment by rotorman** [comment]
> Is there any reason for the power bit to not be the highest, so that it isn't in the middle of the array if more options are added later? Not that it really makes any difference either way.

I can change this to the MSBit if that is desired?

> I think the question should more be "what other transmitters support soft power control of the serial power pins... if the answer is none... the answer is pretty clear ;)

I am presently aware of TX16S only. If there are no other, I will limit the GUI to TX16S only.

**Comment by rotorman** [comment]
> I think the question should more be "what other transmitters support soft power control of the serial power pins... if the answer is none... the answer is pretty clear ;)

I am presently aware of TX16S only, and thus limited the GUI additions in this PR to TX16S.

**Comment by elecpower** [comment]
> Needs Companion support - @elecpower yelp!

@rotorman hear you

File: radio/src/serial.cpp [hunk 2]
```diff
@@ -304,9 +332,8 @@ void serialInit(uint8_t port_nr, int mode)
         state->usart_ctx = port->uart->init(&params);
 
       // Set power on/off
-      if (port->set_pwr) {
-        port->set_pwr(power_required);
-      }
+      if (port_nr < SP_VCP)
```
Comment (pfeerick) [inline radio/src/serial.cpp:335]: If you wrap this with `#if defined(SWSERIALPOWER)` this should implicity optimise `serialSetPowerState`, `getSerialPower`, `setSerialPower` etc out on radios that don't have this... only save about 64 bytes firmware size, but hey, it's a 'free' saving :)

**Comment by pfeerick** [comment]
> I am presently aware of TX16S only, and thus limited the GUI additions in this PR to TX16S.

No, better that, you limited it via a feature flag ... thus it should just be a matter of adding that flag to any other transmitters with soft serial power control ;) 

Changing the bit was really just up for discussion, but thank you :)

File: radio/src/serial.cpp (hunk 2 above)
Comment (rotorman) [reply radio/src/serial.cpp:336]: Thx for the tip, fixed in latest commit.

**Comment by rotorman** [comment]
@elecpower Thank you not only for Companion additions, but also pointing me to the right direction - you were perfectly correct about the issue stemming from power bit position shift - I needed to add 3 YAML_PADDING bits to accommodate the change as well. The PR should be in working condition again.

**Comment by elecpower** [comment]
Confirm power value now interpreted correctly.

However, there is a bug in the simulator which I'm not sure will show up in the real radio. Steps to replicate in Ubuntu 20.04:
1. Companion new file and new model
2. Radio Settings Hardware
3. Set AUX1 to SBUS Trainer and power checked
4. Simulate radio
5. Radio SYS hardware and values match
6. Close sim
7. Companion change AUX1 from SBUS Trainer to OFF and power to unchecked
8. Simulate radio
9. Radio SYS hardware and port has old values

Would appear the memory space is not initialised so could show anything rather than old values as in my example.

@rotorman I've finished my testing so I'll hand back over to you to continue

**Comment by elecpower** [comment]
I noticed during testing that the serial port mode dropdown choices can vary between Companion and the simulator if the libsim has not been compiled with CLI and/or Debug.
Is this an issue, and if, so should Companion radio profile have a couple of Build Options check boxes?

**Comment by raphaelcoeffic** [comment]
> I noticed during testing that the serial port mode dropdown choices can vary between Companion and the simulator if the libsim has not been compiled with CLI and/or Debug. Is this an issue, and if, so should Companion radio profile have a couple of Build Options check boxes?

No, we should assume defaults as used by the GH builds.

**Comment by pfeerick** [comment]
Or rather, perhaps this means `libsim` needs to be compiled with the same options as the radio, if they are missing? ;)

**Comment by pfeerick** [comment]
Preliminary testing suggests there is discrepancy between how Companion and firmware are working... i.e. on the radio, you can toggle the port power regardless of the mode (which IMO is valid as there one it not technically dependant on the other, but this also means you can't use the port for power for *something else* ) but it seems that Companion is only writing the power value when the port has a mode set. 

On a side note, I also notice Companion is not setting the USB-VCP port on TX16S to CLI by default (which is still ok, as the radio will populate this itself if it is missing).

**Comment by rotorman** [comment]
I am in favor of being able to turn on or off AUX[1|2] power, irrespective of the mode selected, and even if the mode is OFF (e.g. to just tap 5V for some DIY mod w/o data lines involved).
@elecpower would you be so kind to match this behaviour in Companion plz as well?

**Comment by pfeerick** [comment]
Radio and companion testing now done... UI is working fine, companion is nuking power state as I expected it would if mode is "OFF". 

An inconsistency in the AUX1 power state - it seems to be ON by default (in contrast to the UI state, and you have to toggle it on and off for it to actually turn off? AUX2 is fine - it is off if set off, and toggles normally. This is also reflected in the bootloader - i.e. AUX1 is on in the bootloader, and AUX2 is off.

**Comment by rotorman** [comment]
Hmm... I wonder if it stems from uninitialized `g_eeGeneral.serialPort` being evaluated during boot/start, as there is no difference in firmware code how AUX1 and AUX2 power is being handled. Need to investigate with ICD.

**Comment by rotorman** [comment]
I had a look at bootloader code and the serial driver is loaded ONLY, if built with DEBUG set:
https://github.com/EdgeTX/edgetx/blob/7a6d6239b5b2fe3efcf1a304602e5839d0b2006e/radio/src/targets/common/arm/stm32/bootloader/boot.cpp#L248-L250
and
https://github.com/EdgeTX/edgetx/blob/7a6d6239b5b2fe3efcf1a304602e5839d0b2006e/radio/src/targets/common/arm/stm32/bootloader/CMakeLists.txt#L141-L149

Hooked up a LA to TX16S and noticed that with 2.7.0-release AUX1 PWR gets ramped up in bootloader as well, whereas AUX2 remains down. Thus, at least the bootloader part seems to be unrelated to this PR.

**Comment by rotorman** [comment]
Found my bug, now the serial power state should come up correctly, irrespective of the mode selected on boot-up. When bootloader is running, serial power is off (except the little glitch on AUX1 during entering bootloader, that happens before the bootloader `main()` runs).

I tested all 4 selection possibilities on TX16S (power for AUX1, AUX2 both off, only AUX1 power on and so on) with a scope. Re-booted also after making a selection to make sure the power is delivered as was selected after restart as well - all seems OK to me now.

**Comment by elecpower** [comment]
> Radio and companion testing now done... UI is working fine, companion is nuking power state as I expected it would if mode is "OFF".

Changed Companion to always save serial port settings

**Review by pfeerick — APPROVED** [review]
Confirmed, power states are behaving exactly as they should for AUX1 and AUX2, on radio. Momentary bootloader 'glitch' of AUX1 power state is perfectly fine. 

I've just pushed a fix for "Telemetry In" not beeing interpreted correctly, but everything else has tested ok, so once a companion test passes it's in. 

image
image

```
TELEMETRY_MIRROR
SBUS_TRAINER
LUA
CLI
GPS
DEBUG
```
were all good.

**Comment by calderra99** [comment]
sorry for jumping in.

i wonder if there's a way that we can assign the function for the UART power availability to a toggle switch? that way we can use itu for an ON-OFF switch for a device attached or drawing power from the said UART, say maybe like an audio Bluetooth module or a telemetry Bluetooth module

**Pull request merged**